Data/ByteString/Builder/RealFloat/Internal.hs

```diff
@@ -0,0 +1,438 @@
+
+{-# INLINABLE (.<<) #-}
+(.<<) :: (Bits a, Integral b) => a -> b -> a
+a .<< s = unsafeShiftL a (fromIntegral s)
+
+{-# INLINABLE mask #-}
+mask :: (Bits a, Integral a) => a -> a
+mask = flip (-) 1 . (.<<) 1
+
+{-# INLINABLE asWord #-}
+asWord :: Integral w => Bool -> w
+asWord = fromIntegral . fromEnum
+
+pmap :: (a -> c) -> (a, b) -> (c, b)
+pmap f (a, b) = (f a, b)
+
+-- Returns the number of decimal digits in v, which must not contain more than 9 digits.
+decimalLength9 :: Word32 -> Int
+decimalLength9 v
+  | v >= 100000000 = 9
+  | v >= 10000000 = 8
+  | v >= 1000000 = 7
+  | v >= 100000 = 6
+  | v >= 10000 = 5
+  | v >= 1000 = 4
+  | v >= 100 = 3
+  | v >= 10 = 2
+  | otherwise = 1
+
+-- Returns the number of decimal digits in v, which must not contain more than 17 digits.
+decimalLength17 :: Word64 -> Int
+decimalLength17 v
+  | v >= 10000000000000000 = 17
+  | v >= 1000000000000000 = 16
+  | v >= 100000000000000 = 15
+  | v >= 10000000000000 = 14
+  | v >= 1000000000000 = 13
+  | v >= 100000000000 = 12
+  | v >= 10000000000 = 11
+  | v >= 1000000000 = 10
+  | v >= 100000000 = 9
+  | v >= 10000000 = 8
+  | v >= 1000000 = 7
+  | v >= 100000 = 6
+  | v >= 10000 = 5
+  | v >= 1000 = 4
+  | v >= 100 = 3
+  | v >= 10 = 2
+  | otherwise = 1
+
+infixr 9 `dot`
+dot :: (a -> b) -> (c -> d -> a) -> c -> d -> b
+dot = (.) . (.)
+
+infixr 9 ...
+(...) :: (a -> b) -> (c -> d -> e -> a) -> c -> d -> e -> b
+(...) = dot . (.)
+
+-- TODO encode proof in types?
+-- From 'In-and-Out Conversions' https://dl.acm.org/citation.cfm?id=362887, we
+-- have that a conversion from a base-b n-digit number to a base-v m-digit
+-- number such that the round-trip conversion is identity requires
+--
+--    v^(m-1) > b^n
+--
+-- Specifically for binary floating point to decimal conversion, we must have
+--
+--    10^(m-1) > 2^n
+-- => log(10^(m-1)) > log(2^n)
+-- => (m-1) * log(10) > n * log(2)
+-- => m-1 > n * log(2) / log(10)
+-- => m-1 >= ceil(n * log(2) / log(10))
+-- => m >= ceil(n * log(2) / log(10)) + 1
+--
+-- And since 32 and 64-bit floats have 23 and 52 bits of mantissa (and then an
+-- implicit leading-bit), we need
+--
+--    ceil(24 * log(2) / log(10)) + 1 => 9
+--    ceil(53 * log(2) / log(10)) + 1 => 17
+--
+-- In addition, the exponent range from floats is [-45,38] and doubles is
+-- [-324,308] (including subnormals) which are 3 and 4 digits respectively
+--
+-- Thus we have,
+--
+--    floats: 1 (sign) + 9 (mantissa) + 1 (.) + 1 (e) + 3 (exponent) = 15
+--    doubles: 1 (sign) + 17 (mantissa) + 1 (.) + 1 (e) + 4 (exponent) = 24
+--
+maxEncodedLength :: Int
+maxEncodedLength = 32
+
+-- TODO TH?
+pokeAll :: String -> Ptr Word8 -> IO (Ptr Word8)
+pokeAll s ptr = foldM pokeOne ptr s
+  where pokeOne p c = S.poke p (c2w c) >> return (p `plusPtr` 1)
+
+boundString :: String -> BoundedPrim ()
+boundString s = boundedPrim maxEncodedLength $ const (pokeAll s)
+
+--         Sign -> Exp  -> Mantissa
+special :: Bool -> Bool -> Bool -> BoundedPrim ()
```

Comment:
Could we possibly define `data Foo = Foo { fooSign :: !Bool, fooExpr :: !Bool, fooMantissa :: !Bool }`, so that the meaning of arguments is self-explanatory, and their ordering is less error-prone?

Reply:
Is https://github.com/haskell/bytestring/pull/365/files#diff-e91607876ac26bb3b25c8b060668d06eb50becfd5a682bd59636e8de5bb9f0d3R214 close enough?